proposals/inline-classes.md

```diff
@@ -0,0 +1,296 @@
+| **Reference type** | Underlying type | Underlying type | Boxed type |
+| **Nullable reference type** | Underlying type  | Boxed type | Boxed type |
+| **Primitive type** | Underlying type  | Boxed type | Boxed type |
+| **Nullable primitive type** | Underlying type  | Boxed type | Boxed type |
+
+For example, from this table it follows that nullable inline class which is based on reference type is mapped to the underlying type:
+```kotlin
+inline class Name(val s: String)
+
+fun foo(n: Name?) { ... } // `Name?` here is mapped to String
+```
+
+Also, if inline class type is used in generic position, then its boxed type will be used:
+```
+// Kotlin: sample.kt
+
+inline class Name(val s: String)
+
+fun generic(names: List<Name>) {} // generic signature will have `List<Name>` as for parameters type
+fun simple(name: Name) {}
+
+// Java
+class Test {
+    void test() {
+        String name = SampleKt.simple();
+        List<Name> ls = Samplekt.generic(); // from Java POV it's List<Name>, not List<String>
+    }
+}
+```
+
+
+#### Generic inline class mapping
+
+Consider the following sample:
+```kotlin
+inline class Generic<T>(val x: T)
+
+fun foo(g: Generic<String>) {}
+```
+
+Now, type `Generic<String>` can be mapped either to `java.lang.String` or to `java.lang.Object`. 
+To make the whole rule more consistent, currently we propose to map `Generic<SomeType>` always to `java.lang.Object`, 
+i.e. we'll map upper bound of type parameter.
+
+* Sidenote: maybe it's worth to consider inline classes with reified generics:
+    ```kotlin
+    inline class Reified<reified T>(val x: T)
+    
+    fun foo(a: Reified<Int>, b: Reified<String>) // a has type `Int`, b has type `String`
+    ``` 
+
+Generic inline classes with underlying value not of type that defined by type parameter are mapped as usual generics:
+```kotlin
+inline class AsList<T>(val ls: List<T>)
+
+fun foo(param: AsList<String>) {}
+```
+
+In JVM signature `param` will have type `java.util.List`, 
+but in generic signature it will be `java.util.List<java.lang.String>` 
+
+## Methods from `kotlin.Any`
+
+Inline classes are indirectly inherited from `Any`, i.e. they can be assigned to a value of type `Any`, but only through boxing.
+
+Methods from `Any` (`toString`, `hashCode`, `equals`) can be useful for a user-defined inline classes and therefore should be customizable. 
+Methods `toString` and `hashCode` can be overridden as usual methods from `Any`. For method `equals` we're going to introduce new operator 
+that represents "typed" `equals` to avoid boxing for inline classes:
+```kotlin
+inline class Foo(val s: String) {
+    operator fun equals(other: Foo): Boolean { ... }
+}
+```
+
+Compiler will generate original `equals` method that is delegated to the typed version.  
+
+By default, compiler will automatically generate `equals`, `hashCode` and `toString` same as for data classes.
```

Comment:
There is an interesting conflict with this section about equals/hashCode and implementing interfaces. 

Let me quote:

>Compiler will generate original `equals` method that is delegated to the typed version. 

and

>Inline class can implement only interfaces

Which means that comparing two *different* inline classes that use same underlying type and implement same interface can result in wrong behavior because of comparing underlying values which can be indeed equal.

Example:

```kotlin
interface Bytable {
    fun toBytes(): ByteArray
}

inline class UserId(private val id: String): Bytable {
    override fun toBytes(): ByteArray = "userId:$id".toByteArray()
}

inline class OrderId(private val id: String) : Bytable {
    override fun toBytes(): ByteArray = "orderId:$id".toByteArray()
}

fun f(bytable1: Bytable, bytable2: Bytable) {
    assert(bytable1 == bytable2)
}

fun test() {
    f(bytable1 = UserId("123"), bytable2 = OrderId("123"))
}

fun main(args: Array<String>) {
    test()
}
```

If you change `inline` classes to `data` classes, test will fail.
But I think it'll pass with `inline` classes which is interesting.

Reply:
I'm intrigued as to how exactly unsigned types will work under this proposal.

I can see how one could do something like this:

```kotlin
val u = UInt(42)
```

but how then would you access the `value` property if it's private or is the idea that `u` itself would be treated as the value?

Also, if you wanted to create a UInt that was greater than Int.MAX_VALUE what value would you present to the primary constructor? Presumably one would not be expected to work out the appropriate negative Int value corresponding to this so would Int literals be extended to permit higher values up to UInt.MAX_VALUE?

Reply:
I don't think number literals for unsigned values are part of this proposal, but I guess if they are implemented they need to allow for higher values than there signed parts. How else are you gonna get values for `ULong`. For `UInt` you could use a function converting a `long`. That being said, maybe you are doing something wrong if you have to hardcode numbers that big.